### PR TITLE
Multiplying by the identity gate should return the other operation

### LINKED ISSUE #1676: Multiplying by the identity gate should return the other operation
```python
a, b, c = cirq.LineQubit.range(3)
assert cirq.X(a) * cirq.I(a) == cirq.X(a)
assert cirq.X(a) * cirq.I(b) == cirq.X(a)
assert cirq.X(a) * cirq.Y(b) * cirq.I(c) == cirq.X(a) * cirq.Y(b)
```

### PR DESCRIPTION
**Pull request by timothyman**
Fixes #1676 
- Added __mul__ and __rmul__ functions in GateOperation and added tests
- It seemed intuitive to me that the multiplication with Identity Gate is commutative, but intuition can be misleading in QM. Please let me know if that's wrong.

### CODE REVIEW

File: cirq/ops/gate_operation.py [hunk 1]
```diff
@@ -186,6 +186,19 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
                               qubits=self.qubits,
                               default=None)
 
+    def __mul__(self, other):
+        from cirq.ops.pauli_string import (SingleQubitPauliStringGateOperation,
```
Comment (vtomole) [inline cirq/ops/gate_operation.py]: Move imports to the top of the file.

File: cirq/ops/pauli_gates_test.py [hunk 2]
```diff
@@ -193,3 +193,13 @@ def test_apply_unitary():
...
+
+
+def test_identity_multiplication():
+    a, b, c = cirq.LineQubit.range(3)
+    assert cirq.X(a) * cirq.I(a) == cirq.X(a)
+    assert cirq.X(a) * cirq.I(b) == cirq.X(a)
+    assert cirq.X(a) * cirq.Y(b) * cirq.I(c) == cirq.X(a) * cirq.Y(b)
+    assert cirq.I(c) * cirq.X(a) * cirq.Y(b) == cirq.X(a) * cirq.Y(b)
+    with pytest.raises(TypeError):
+        _ = cirq.H(c) * cirq.X(a) * cirq.Y(b)
```
Comment (vtomole) [inline cirq/ops/pauli_gates_test.py]: Add newline.

Comment (timothyman) [reply cirq/ops/pauli_gates_test.py]: Originally I did, but then the formatter complained.

> Changes in cirq/ops/pauli_gates_test.py require formatting:
> --- cirq/ops/pauli_gates_test.py	(original)
> +++ cirq/ops/pauli_gates_test.py	(reformatted)
> @@ -203,4 +203,3 @@
>      assert cirq.I(c) * cirq.X(a) * cirq.Y(b) == cirq.X(a) * cirq.Y(b)
>      with pytest.raises(TypeError):
>          cirq.H(c) * cirq.X(a) * cirq.Y(b)
> -        
> Some formatting needed on changed lines.
> The command "check/format-incremental" exited with 1.

File: cirq/ops/gate_operation.py (hunk 1 above)
Comment (timothyman) [reply cirq/ops/gate_operation.py]: I believe this is a "hack" to avoid circular dependencies/imports?

See for instance:
https://github.com/quantumlib/Cirq/blob/28a575d332573d4ddd4d056058caf251a06939bb/cirq/ops/pauli_string.py#L120

But to be honest, I haven't tried moving the imports to the top of the file.

Comment (timothyman) [reply cirq/ops/gate_operation.py]: I have tried it, and indeed everything fails after I moved the imports to the top of the file

**Comment by timothyman** [comment]
Thanks for the feedback and the design. That is indeed a lot better design.

File: cirq/ops/gate_operation.py [hunk 3]
```diff
@@ -186,13 +186,19 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
                               qubits=self.qubits,
                               default=None)
 
+
+class IdentityOperation(GateOperation):
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Inherit from `Operation`, not `GateOperation`.

Comment (Strilanc) [reply cirq/ops/gate_operation.py]: Rule of thumb: whenever you find yourself throwing exceptions like "I know I said I'm an X but I don't do the thing Y that Xs normally do", you should not be declaring yourself to be an X.

Comment (timothyman) [reply cirq/ops/gate_operation.py]: Again, fair criticism, thanks. I was just surprised how much duplicate code from GateOperation I had to introduce in the IdentityOperation when inheriting from Operation to make it past the older tests. Methods like `__init__`, `__repr__`, `__str__`, `_pauli_expansion_` and more are exactly the same. That didn't look good to me either.

Let me try again and let you be the judge.

File: cirq/ops/gate_operation.py [hunk 4]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
                               default=None)
 
 
+@value.value_equality(approximate=True)
+class IdentityOperation(raw_types.Operation):
+    """An application of the identity gate to a sequence of qubits."""
+
+    def __init__(self, gate: common_gates.IdentityGate,
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Drop the gate parameter.

File: cirq/ops/gate_operation.py [hunk 5]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+                 qubits: Sequence[raw_types.Qid]) -> None:
+        """
+        Args:
+            gate: The gate to apply. Only really accepts the IdentityGate
+            qubits: The qubits to operate on.
+        """
+        gate.validate_args(qubits)
+        self._gate = gate
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Drop the `_gate` attribute.

File: cirq/ops/gate_operation.py [hunk 6]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+    """An application of the identity gate to a sequence of qubits."""
+
+    def __init__(self, gate: common_gates.IdentityGate,
+                 qubits: Sequence[raw_types.Qid]) -> None:
+        """
+        Args:
+            gate: The gate to apply. Only really accepts the IdentityGate
+            qubits: The qubits to operate on.
+        """
+        gate.validate_args(qubits)
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Any set of qubits will be acceptable.

File: cirq/ops/gate_operation.py [hunk 7]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+        self._gate = gate
+        self._qubits = tuple(qubits)
+
+    @property
+    def gate(self) -> raw_types.Gate:
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Drop this getter.

File: cirq/ops/gate_operation.py [hunk 8]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+        """The gate applied by the operation."""
+        return self._gate
+
+    @property
+    def qubits(self) -> Tuple[raw_types.Qid, ...]:
+        """The qubits targeted by the operation."""
+        return self._qubits
+
+    def with_qubits(self, *new_qubits: raw_types.Qid) -> 'raw_types.Operation':
+        return self.gate.on(*new_qubits)
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: `return IdentityOperation(*new_qubits)`

File: cirq/ops/gate_operation.py [hunk 9]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+
+    def __repr__(self):
+        # Abbreviate when possible.
+        if self == self.gate.on(*self.qubits):
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]:
```
if len(self.qubits) == 1:
    return f'cirq.I.on({self.qubits[0]})'
```

File: cirq/ops/gate_operation.py [hunk 10]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+            return '{!r}.on({})'.format(self.gate,
+                                        ', '.join(repr(q) for q in self.qubits))
+
+        return 'cirq.IdentityOperation(gate={!r}, qubits={!r})'.format(
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: `return 'cirq.IdentityOperation({!r})'.format(list(self.qubits))`

File: cirq/ops/gate_operation.py [hunk 11]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+
+    def __str__(self):
+        return '{}({})'.format(self.gate,
+                               ', '.join(str(e) for e in self.qubits))
+
+    def _group_interchangeable_qubits(
+            self
+    ) -> Tuple[Union[raw_types.Qid, Tuple[int, FrozenSet[raw_types.Qid]]], ...]:
+
+        if not isinstance(self.gate, gate_features.InterchangeableQubitsGate):
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Delete this method; we know the qubits are interchangeable.

File: cirq/ops/gate_operation.py [hunk 12]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+        groups = {}  # type: Dict[int, List[raw_types.Qid]]
+        for i, q in enumerate(self.qubits):
+            k = self.gate.qubit_index_to_equivalence_group_key(i)
+            if k not in groups:
+                groups[k] = []
+            groups[k].append(q)
+        return tuple(sorted((k, frozenset(v)) for k, v in groups.items()))
+
+    def _value_equality_values_(self):
+        return self.gate, self._group_interchangeable_qubits()
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: `return frozenset(self.qubits)`

File: cirq/ops/gate_operation.py [hunk 13]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+
+    def _pauli_expansion_(self) -> value.LinearDict[str]:
+        return protocols.pauli_expansion(self.gate)
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: `return value.LinearDict({'I' * len(self.qubits): 1.0})`

File: cirq/ops/gate_operation.py [hunk 14]
```diff
@@ -187,6 +187,85 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+
+    def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs
+                       ) -> Optional[np.ndarray]:
+        return args.target_tensor
+
+    def _circuit_diagram_info_(self, args: protocols.CircuitDiagramInfoArgs
+                              ) -> protocols.CircuitDiagramInfo:
+        return protocols.circuit_diagram_info(self.gate, args, NotImplemented)
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: `return ('I',) * len(self.qubits)`

**Comment by timothyman** [comment]
@Strilanc Ready for re-review.

Sorry for the back and forth.

File: cirq/ops/common_gates.py [hunk 15]
```diff
@@ -627,6 +627,16 @@ def _circuit_diagram_info_(self,
...
 
+    def on(self, *qubits: raw_types.Qid) -> raw_types.Operation:
+        """Returns an application of this gate to the given qubits.
+
+        Args:
+            *qubits: The collection of qubits to potentially apply the gate to.
+        """
+        # Avoids circular import.
+        from cirq.ops import gate_operation
+        return gate_operation.IdentityOperation(list(qubits))
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: Move `IdentityGate` and `IdentityOperation` into a file `cirq/ops/identity.py`.

File: cirq/ops/gate_operation.py [hunk 16]
```diff
@@ -187,6 +187,66 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+        """
+        Args:
+            qubits: The qubits to operate on.
+        """
+        if len(qubits) == 0:
+            raise ValueError(
+                'Applied an identity gate to an empty set of qubits.')
+
+        if any([not isinstance(qubit, raw_types.Qid) for qubit in qubits]):
+            raise ValueError('Gate was called with type different than Qid.')
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: `raise ValueError('Gave non-Qid objects to IdentityOperation: {!r}'.format(qubits))`

File: cirq/ops/gate_operation.py [hunk 17]
```diff
@@ -187,6 +187,66 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+    def __init__(self, qubits: Sequence[raw_types.Qid]) -> None:
+        """
+        Args:
+            qubits: The qubits to operate on.
+        """
+        if len(qubits) == 0:
+            raise ValueError(
+                'Applied an identity gate to an empty set of qubits.')
+
+        if any([not isinstance(qubit, raw_types.Qid) for qubit in qubits]):
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Unnecessary square brackets

File: cirq/ops/gate_operation.py [hunk 18]
```diff
@@ -187,6 +187,66 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+        """The qubits targeted by the operation."""
+        return self._qubits
+
+    def with_qubits(self, *new_qubits: raw_types.Qid) -> 'raw_types.Operation':
+        return IdentityOperation(new_qubits)
+
+    def __repr__(self):
+        # Abbreviate when possible.
+        if len(self.qubits) == 1:
+            return f'cirq.I.on({self._qubits[0]})'
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: This is not correct; it's using the str instead of the repr of the qubit. I think you need `:!r` somewhere.

File: cirq/ops/gate_operation.py [hunk 19]
```diff
@@ -187,6 +187,66 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Add a test that calls `cirq.testing.assert_implements_consistent_protocols` on a single-qubit and also a multi-qubit IdentityOperation instance.

File: cirq/ops/gate_operation.py [hunk 20]
```diff
@@ -187,6 +187,66 @@ def _qasm_(self, args: protocols.QasmArgs) -> Optional[str]:
...
+        return frozenset(self._qubits)
+
+    def _pauli_expansion_(self) -> value.LinearDict[str]:
+        return value.LinearDict({'I' * len(self._qubits): 1.0})
+
+    def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs
+                       ) -> Optional[np.ndarray]:
+        return args.target_tensor
+
+    def _circuit_diagram_info_(self, args: protocols.CircuitDiagramInfoArgs
```
Comment (Strilanc) [inline cirq/ops/gate_operation.py]: Add a `__pow__` method that returns self when `isinstance(power, (int, float, complex, sympy.Basic))`, else NotImplemented.

File: cirq/ops/gate_operation_test.py [hunk 21]
```diff
@@ -240,6 +249,11 @@ def test_repr():
     assert repr(cirq.GateOperation(cirq.CZ, (a, b))
                 ) == 'cirq.CZ.on(cirq.LineQubit(0), cirq.LineQubit(1))'
 
+    assert repr(cirq.IdentityOperation((a,))) == ('cirq.I.on(0)')
```
Comment (Strilanc) [inline cirq/ops/gate_operation_test.py]: Bad test. Should be `cirq.I.on(cirq.LineQubit(0))`.

**Comment by Strilanc** [comment]
Add `def _trace_distance_bound_(self): return 0` to the class

**Comment by timothyman** [comment]
@Strilanc Ready for re-review.

**Comment by timothyman** [comment]
@Strilanc What's the status of this? Still needed?

**Comment by vtomole** [comment]
Hey @timothyman , I'd like to get this into master. Please fix the merge conflicts and we should be good to go.

File: cirq/ops/pauli_gates_test.py [hunk 22]
```diff
@@ -198,6 +198,20 @@ def test_apply_unitary():
     cirq.testing.assert_has_consistent_apply_unitary(cirq.Z)
 
 
+def test_identity_multiplication():
+    a, b, c = cirq.LineQubit.range(3)
+    assert cirq.X(a) * cirq.I(a) == cirq.X(a)
+    assert cirq.X(a) * cirq.I(b) == cirq.X(a)
+    assert cirq.X(a) * cirq.Y(b) * cirq.I(c) == cirq.X(a) * cirq.Y(b)
+    assert cirq.I(c) * cirq.X(a) * cirq.Y(b) == cirq.X(a) * cirq.Y(b)
```
Comment (Strilanc) [inline cirq/ops/pauli_gates_test.py]: Check some non-pauli single qubit gates and some multi-qubit gates.

**Comment by timothyman** [comment]
I had to rename the helper function from identity to identity_each, so make it unambiguous what you're importing. Please let me know if you prefer another name or route.

File: cirq/ops/pauli_string.py [hunk 23]
```diff
@@ -24,8 +24,16 @@
...
-                      pauli_gates, clifford_gate, pauli_interaction_gate)
+from cirq.ops import (
+    global_phase_op,
+    raw_types,
+    gate_operation,
+    common_gates,
+    pauli_gates,
+    clifford_gate,
+    pauli_interaction_gate,
+    identity as identity,
```
Comment (vtomole) [inline cirq/ops/pauli_string.py]: Is this redundant?

File: cirq/ops/identity_test.py [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+
+def test_identity_global():
+    qubits = cirq.LineQubit.range(3)
+    assert cirq.identity_each(*qubits) == cirq.IdentityGate(3).on(*qubits)
+    qids = cirq.LineQid.for_qid_shape((1, 2, 3))
+    assert cirq.identity_each(*qids) == cirq.IdentityGate(3,
+                                                          (1, 2, 3)).on(*qids)
+    with pytest.raises(ValueError, match='type different'):
+        cirq.identity_each(
+            qubits)  # The user forgot to expand the list for example.
```
Comment (vtomole) [inline cirq/ops/identity_test.py]: Add newline.

File: cirq/ops/identity.py [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+        An identity operation on the given qubits.
+
+    Raises:
+        ValueError if the qubits are not instances of Qid.
+    """
+    if not all(isinstance(qubit, raw_types.Qid) for qubit in qubits):
+        raise ValueError('identity() was called with type different than Qid.')
+
+    qid_shape = protocols.qid_shape(qubits)
+    return IdentityGate(len(qubits), qid_shape).on(*qubits)
```
Comment (vtomole) [inline cirq/ops/identity.py]: Add newline.

File: cirq/ops/linear_combinations.py [hunk 26]
```diff
@@ -107,22 +107,22 @@ def __pow__(self, exponent: int) -> 'LinearCombinationOfGates':
         if self.num_qubits() != 1:
             return NotImplemented
         pauli_basis = {
-            common_gates.I,
+            I,
```
Comment (vtomole) [inline cirq/ops/linear_combinations.py]: Use `identity.I` as you have been elsewhere.

File: cirq/ops/pauli_string.py (hunk 23 above)
Comment (timothyman) [reply cirq/ops/pauli_string.py]: Yes. Removed.

File: cirq/ops/identity_test.py (hunk 24 above)
Comment (timothyman) [reply cirq/ops/identity_test.py]: Added.

File: cirq/ops/identity.py (hunk 25 above)
Comment (timothyman) [reply cirq/ops/identity.py]: Added.

File: cirq/ops/linear_combinations.py (hunk 26 above)
Comment (timothyman) [reply cirq/ops/linear_combinations.py]: Good find! Resolved. Thanks for the review.

File: cirq/ops/pauli_gates.py [hunk 27]
```diff
@@ -15,7 +15,7 @@
 from typing import Union, TYPE_CHECKING, Tuple, cast
 
 from cirq import value
-from cirq.ops import common_gates, raw_types
+from cirq.ops import common_gates, raw_types, identity as identity
```
Comment (vtomole) [inline cirq/ops/pauli_gates.py]: Here too.

Comment (vtomole) [reply cirq/ops/pauli_gates.py]: @timothyman This hasn't been resolved yet because ` identity as identity` is still here.

Comment (timothyman) [reply cirq/ops/pauli_gates.py]: @vtomole Are you sure? Not on my branch, and not on the files changed here.

Comment (vtomole) [reply cirq/ops/pauli_gates.py]: I've had this happen to me where the diff doesn't change. Maybe its a caching thing. Anyway, this all looks good to me.
